cache/expirationcache/prefetching_cache.go

```diff
@@ -0,0 +1,125 @@
+package expirationcache
+
+import (
+	"time"
+)
+
+type PrefetchingExpiringLRUCache[T any] struct {
+	cache                   ExpiringCache[cacheValue[T]]
+	prefetchingNameCache    ExpiringCache[int]
+	reloadFn                ReloadEntryFn[T]
+	prefetchThreshold       int
+	prefetchExpires         time.Duration
+	onPrefetchEntryReloaded OnEntryReloadedCallback
+	onPrefetchCacheHit      OnCacheHitCallback
+}
+
+type cacheValue[T any] struct {
+	element  *T
+	prefetch bool
+}
+
+// OnEntryReloadedCallback will be called if a prefetched entry is reloaded
+type OnEntryReloadedCallback func(key string)
+
+// ReloadEntryFn reloads a prefetched entry by key
+type ReloadEntryFn[T any] func(key string) (*T, time.Duration)
+
+type PrefetchingOptions[T any] struct {
+	Options
+	ReloadFn                func(cacheKey string) (*T, time.Duration)
+	PrefetchThreshold       int
+	PrefetchExpires         time.Duration
+	PrefetchMaxItemsCount   int
+	OnPrefetchAfterPut      OnAfterPutCallback
+	OnPrefetchEntryReloaded OnEntryReloadedCallback
+	OnPrefetchCacheHit      OnCacheHitCallback
+}
+
+type PrefetchingCacheOption[T any] func(c *PrefetchingExpiringLRUCache[cacheValue[T]])
+
+func NewPrefetchingCache[T any](options PrefetchingOptions[T]) *PrefetchingExpiringLRUCache[T] {
+	pc := &PrefetchingExpiringLRUCache[T]{
+		prefetchingNameCache: NewCache[int](Options{
+			CleanupInterval: time.Minute,
+			MaxSize:         uint(options.PrefetchMaxItemsCount),
+			OnAfterPutFn:    options.OnPrefetchAfterPut,
+		}),
+		prefetchExpires:         options.PrefetchExpires,
+		prefetchThreshold:       options.PrefetchThreshold,
+		reloadFn:                options.ReloadFn,
+		onPrefetchEntryReloaded: options.OnPrefetchEntryReloaded,
+		onPrefetchCacheHit:      options.OnPrefetchCacheHit,
+	}
+
+	pc.cache = NewCacheWithOnExpired[cacheValue[T]](options.Options, pc.onExpired)
+
+	return pc
+}
+
+// check if a cache entry should be prefetched: was queried > threshold in the time window
+func (e *PrefetchingExpiringLRUCache[T]) shouldPrefetch(cacheKey string) bool {
+	if e.prefetchThreshold == 0 {
+		return true
+	}
+
+	cnt, _ := e.prefetchingNameCache.Get(cacheKey)
+
+	return cnt != nil && *cnt > e.prefetchThreshold
+}
+
+func (e *PrefetchingExpiringLRUCache[T]) onExpired(cacheKey string) (val *cacheValue[T], ttl time.Duration) {
+	if e.shouldPrefetch(cacheKey) {
+		loadedVal, ttl := e.reloadFn(cacheKey)
+		if loadedVal != nil {
+			if e.onPrefetchEntryReloaded != nil {
+				e.onPrefetchEntryReloaded(cacheKey)
+			}
+
+			return &cacheValue[T]{loadedVal, true}, ttl
+		}
+	}
+
+	return nil, 0
+}
+
+func (e *PrefetchingExpiringLRUCache[T]) trackCacheKeyQueryCount(cacheKey string) {
+	var count int
+	if x, _ := e.prefetchingNameCache.Get(cacheKey); x != nil {
+		count = *x
+	}
+	count++
+	e.prefetchingNameCache.Put(cacheKey, &count, e.prefetchExpires)
```

Comment:
yes, I was thinking about it some time ago. It would be more flexible, so you can always decide: use reference or a value. But on the other side: using values in a cache can lead in performance issues since it will create a copy each time.

In this case (prefetching), maybe we should use a custom type instead of int and just call "increment" as method?

Reply:
I changed int to atomic.Uint32. Imho it looks better now, no referencing/dereferencing to primitive type necessary.